### PR TITLE
add hdf5 reader

### PR DESCRIPTION
**Pull request by wangzhezhe**
1 add hdf5 reader which load the image array and generate the Block
2 the information to generate Header in Block should be set manually
3 tested ok for big_stem_image.h5

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -32,14 +32,16 @@ include_directories(
 find_package(Eigen3 REQUIRED)
 include_directories(${EIGEN3_INCLUDE_DIR})
 
+find_package(pybind11 REQUIRED)
+
```
Comment (cjh1) [inline CMakeLists.txt]: As we are doing a find here, we should remove the extra call that is being done in the `/python/CMakeLists.txt`

File: examples/hdf5Reader/reader_hdf5.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    img.save(name)
+
+# init reader
+filename="./big_stem_image.h5"
+h5file= h5py.File(filename, 'r')
+
+inner_radii = [0, 40]
+outer_radii = [288, 288]
+
+imgs = image.create_stem_images(h5file, inner_radii, outer_radii, width=160, height=160)
```
Comment (cjh1) [inline examples/hdf5Reader/reader_hdf5.py]: Nice, this look much cleaner.

File: python/reader_h5.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    m_blockNumInFile(blockNumInFile), m_totalImageNum(totalImageNum),
+    m_imageNumbers(imageNumbers)
+{}
+
+Block H5Reader::read()
+{
+
+  // get to the end of the file, return empty Block
+  if (m_currIndex >= m_totalImageNum) {
+    std::cout << "data loading finish with index " << m_currIndex << std::endl;
```
Comment (cjh1) [inline python/reader_h5.cpp]: We should remove this debug.

File: python/reader_h5.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  py::buffer_info buf = pyarray.request();
+  uint32_t arraySize = pyarray.size();
+
+  uint16_t* ptr = (uint16_t*)buf.ptr;
+
+  Header h = Header(m_imageWidth, m_imageHeight, m_imageNumInBlock,
+                    m_blockNumInFile, m_scanWidth, m_scanHeight, m_currIndex, m_imageNumbers);
+  Block b = Block(h);
+
+  std::copy(ptr, ptr + arraySize, b.data.get());
```
Comment (cjh1) [inline python/reader_h5.cpp]: Why are we doing this copy?

File: python/reader_h5.h [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+    iterator(H5Reader* h5reader) : m_H5Reader(h5reader)
+    {
+      if(h5reader==nullptr){
+        return;
+      }
+      // read data at first time
+      m_block = m_H5Reader->read();
+      if (!m_block.data) {
+        std::cout<<"m_H5Reader is null" << std::endl;
```
Comment (cjh1) [inline python/reader_h5.h]: debug?

File: python/stempy/image/__init__.py [hunk 6]
```diff
@@ -1,9 +1,40 @@
 from stempy import _image
+from stempy import _io
 import numpy as np
-
-def create_stem_images(reader, inner_radii,
+import h5py
+
+
+def get_hdf5_reader(h5file):
```
Comment (cjh1) [inline python/stempy/image/__init__.py]: Can we add a leading underscore to indicate that this is an internal function

File: python/stempy/image/__init__.py [hunk 7]
```diff
@@ -1,9 +1,40 @@
...
+    # the initialization is at the io.cpp
+    dset_frame=h5file['frames']
+    dset_frame_shape=dset_frame.shape
+    totalImgNum=dset_frame_shape[0]
+    imagewidth=dset_frame_shape[1]
+    imageheight=dset_frame_shape[2]
+    
+    dset_stem_shape=h5file['stem/images'].shape
+    scanwidth=dset_stem_shape[1]
+    scanheight=dset_stem_shape[2]
```
Comment (cjh1) [inline python/stempy/image/__init__.py]: `stem/images` is unlikely to be present if we are trying to create a STEM image. We need to look for attributes `Nx` and `Ny`, that we can use to save the scan dimension in.

File: python/stempy/image/__init__.py [hunk 8]
```diff
@@ -1,9 +1,40 @@
...
+    
+    # the number here are same with the smallscan data
+    blocksize=32
+    blockNumInFile=25
```
Comment (cjh1) [inline python/stempy/image/__init__.py]: If these are constants we should move them to the top of the file and use uppercase.

File: python/stempy/image/__init__.py [hunk 9]
```diff
@@ -1,9 +1,40 @@
...
+    image_numbers = []
+    
+    h5reader = _io._h5reader(dset_frame, image_numbers, imagewidth, imageheight, scanwidth, scanheight, blocksize, blockNumInFile, totalImgNum)
+    return h5reader
+
+def create_stem_images(input, inner_radii,
                        outer_radii, width=0, height=0,
                        center_x=-1, center_y=-1):
+    # check if the input is the hdf5 dataset
+    if(isinstance(input, h5py._hl.files.File)):
```
Comment (cjh1) [inline python/stempy/image/__init__.py]: I think we can broaden this check to `hasattr(input, '__len__') and hasattr(input, '__getitem__')`, then is should also work with regular numpy arrays. We should then remain some the of the classes so they don't explicitly refer to HD5, for example `H5Reader` would become `SequenceReader`.

**Comment by cjh1** [comment]
Looks like you need merge in master as well.

**Comment by cjh1** [comment]
@alesgenova This is the HDF5 PR

**Comment by alesgenova** [comment]
@wangzhezhe 
I'm trying to generate a stem image from a hdf5 file, but I'm running into issues. Not sure if the problem is in my code or in stempy.

Here is an example:
```python
import h5py
from stempy import io, image

filename = '/home/alessandro/Downloads/stem_image.h5'

h5file= h5py.File(filename, 'r')

inner_radii = (0,)
outer_radii = (40,)
imgs = image.create_stem_images(h5file, inner_radii, outer_radii, width=160, height=160)
```

It produces this debugging output when run
```
hdf5 reader is used
debug constructor
986
998
839
1077
902
1012
979
1027
1094
970
construct pyblock with index 0and 32
debug getData
pyblock get data is called 
pyblock ok to return shared pointer 
986
998
839
1077
902
1012
979
1027
1094
970
return block
m_H5Reader is null
```

And it crashes with the following trace
```
Traceback (most recent call last):
  File "test.py", line 11, in <module>
    imgs = image.create_stem_images(h5file, inner_radii, outer_radii, width=160, height=160)
  File "/home/alessandro/kitware/stempy-build/lib/stempy/image/__init__.py", line 40, in create_stem_images
    center_x, center_y)
ValueError: No blocks to read!
```

**Comment by wangzhezhe** [comment]
Hello, @alesgenova  This branch is on progress and there are some bugs need to be fixed when we update the block into more efficient one. Please check this branch https://github.com/wangzhezhe/stempy/tree/feature/add_hdf5_reader_withoriginalblock , thanks!

**Pull request closed without merging**